### ISSUE TITLE
New property: rulers

### ISSUE THREAD

**Issue by zimbatm** [editorconfig/editorconfig#89]
Visual cues to significant vertical positions in the code
- rulers: array of int where int is a column number

By default `rulers` are not set. Setting `rulers=0` is equivalent to not setting any rulers.
### Example

```ini
[*.js]
rulers = 80, 120
```
### Rationale

You want to be able to impose a width limit on some files. For example git commits are traditionally limited to 50 characters. I use a 79 width on Markdown files to avoid horizontal scrolling and keep the space for side-by-side comparisons.

Because common behaviour is hard to specialise the indicators are only visual.
### Editor behaviour

The editor should display vertical rulers at the given rulers columns. It is also important that the user can override or turn this feature off if desired so.
### Possible expansions

The last column could optionally be used as a hard-wrapping cue.

[Wiki page for dumping research on support in different editors](https://github.com/editorconfig/editorconfig/wiki/Property-research:-maximum-line-length)
##

**Comment by rayshan** [comment]
:+1:

**Comment by xuhdev** [comment]
Currently a similar feature [max_line_length](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#max_line_length) is implemented in limited number of editors. You may check it out.

**Comment by rayshan** [comment]
Ahh I was only looking at the website. Thanks!

**Comment by ljani** [comment]
What's the preferred way to force disable this? `max_line_length` spec says only positive integers are allowed. If it doesn't exist, should the plugin disable it?

I'm asking because I cannot disable `textwidth=80` inherited from my `.vimrc` with [editorconfig-vim](https://github.com/editorconfig/editorconfig-vim). Currently as a workaround, I use `max_line_length = 1024`.

**Comment by zimbatm** [comment]
@xuhdev `max_line_length` is not what this proposal is focusing on because it has too many corner-cases. We just want a ruler to signal desired boundaries.

@ljani `textwidth=0` should disable that setting

**Comment by xuhdev** [comment]
@zimbatm You're right, they are not the same. Just to mention some people would simply use `max_line_length` to solve their problems.

**Comment by ljani** [comment]
@zimbatm Yes, it works very well, if I set `textwidth=0` in Vim, but I can't use editorconfig to do this, can I? I mean I need to override whatever the editor might define and disable it.

Also, I noticed this seems to be the wrong ticket, sorry :)

**Comment by zimbatm** [comment]
If only I had the energy to convert this proposal into an actual spec... :)

**Comment by tregusti** [comment]
If only =)

**Comment by ulope** [comment]
Is there anything we (i.e. the community) can do to help with this?

For me line length is the second most important setting that should be consistent across a team.

**Comment by tregusti** [comment]
Thinking out loud:

```
# visual indicator at 80, forces new line at 100
x: 80, 100 hard
# same as above
x: 80 soft, 100 hard
# does not force anything just pretty lines
x: 80, 100, 120, 140, 200
```

or maybe:

```
# when passing 80, lines turns red or something decided by editor/theme
x: 80 warn, 100 hard
x:

**Comment by demiters** [comment]
How should one go about setting this for Sublime Text 3? Or is it still not supported?

**Comment by xuhdev** [comment]
@sindresorhus Do you want to implement this feature with `max_line_length` in sublime text 3?

**Comment by sindresorhus** [comment]
@xuhdev I think hard-wrapping is archaic and pointless today when all editors supports soft-wrapping, so won't be done by me, but if someone does I good pull request I would accept it.

**Comment by jedmao** [comment]
@sindresorhus it's true that editors support soft-wrapping, but don't you think hard-wrapping would still be valuable in diffs (e.g., GitHub)?

**Comment by sindresorhus** [comment]
@jedmao GitHub supports soft-wrap in diffs.

**Comment by jedmao** [comment]
@sindresorhus perhaps [on prose documents](https://github.com/blog/1707-soft-wrapping-on-prose-diffs), but I know I've had to scroll to the right recently on code files; specifically, JSX files.

**Comment by xuhdev** [comment]
@sindresorhus On text files, soft wraps have a few disadvantages:

- Hard to apply small patches.
- Hard to blame.
- In the context of indent based syntax, it's hard to see the indentation.

Any way, I'm not saying that soft wraps aren't useful; I just want to show you that hard wraps have some values in it and thus it makes sense to support this feature (I don't use tabs for indentation but it does not mean I don't support it because many other people use it).

**Comment by Chiel92** [comment]
Just another use case.
For vim, the plugin [vim-autoformat](https://github.com/Chiel92/vim-autoformat) uses the textwidth property by passing it on to code formatters. Many code formatters will try to reformat the code to make all lines be smaller than the maximum width. I can imagine that formatter integration plugins in other editors do the same.

**Comment by terrycloth** [comment]
Atom's EditorConfig plugin seems to have implemented `max_line_length` as equivalent to what OP is asking for --- a visual hint rather than forcing a hard wrap --- and requires the other Atom package wrap-guide to be installed and enabled.

**Comment by Zastai** [comment]
Any chance the IntelliJ platform (basis of the JetBrains IDEs like IDEA, WebStorm, etc.) could be added to the editor research item?
They support `max_line_length` (both showing a visual guide and setting line wrapping at that column, depending on wrapping-related style configuration).

As for a setting to configure visual guides (separately; the default could be "whatever is specified via `max_line_length`"), an important use case would be older languages like COBOL, which are column-based.
For example, for COBOL I would want `max_line_length = 80`, plus something like `rulers = 7, 8, 12, 73` (1-6 = sequence number area, 7 = indicator area, 8-11 = area A, 12-72 = area B, 73-end = comment area; see also [the IBM COBOL compiler manual](https://www.ibm.com/support/knowledgecenter/SS6SG3_6.2.0/com.ibm.cobol62.ent.doc/PGandLR/ref/rlfmt.html)).

**Comment by dyadix** [comment]
@Zastai I'm not sure I understand what research is need for IntelliJ platform. If needed, the property will be added there. There is already a similar settings but not accessible via EditorConfig yet. The only question is, can it be added by `rulers` name or should it be a domain-specific property if `rulers` name is not approved yet and can change later.

**Comment by dyadix** [comment]
Is there any decision about `rulers` property? I need either to use a proprietary `ij_visual_guides` for Intellij IDEA or use a standard name, e.g. `rulers`. Of course a support for a standard name can be added later but it may somewhat confuse end users since both names will co-exist for a while.

**Comment by jedmao** [comment]
@dyadix the short answer is no. I would love to see some resolution here though. See #387

**Comment by ssg** [comment]
Visual Studio Editor Guidelines extensions already implements a `guidelines` property for the same purpose: https://marketplace.visualstudio.com/items?itemName=PaulHarrington.EditorGuidelines

**Comment by WhyNotHugo** [comment]
I've updated the wiki page with further details on this topic.

It seems the wiki page only covered the behaviour mentioned in comments (hard-wrapping), but not the actual topic of this issue: adding vertical rules.

I've made sure both items are correctly mentioned in the wiki now.

---

With the given amount of research, what's the status of this issue? Is further consensus required, or are we just waiting for a PR?

**Comment by justinmcgrath** [comment]
If wrapping is implemented, could there be separate options for block comments, inline comments and code? A narrow column limit for code is often not sensible with modern displays, but it is still nice for block comments, which are usually written as paragraphs. Using a 140 column limit, for example, is often reasonable for code, but makes for badly formatted block comments.

**Comment by toolforger** [comment]
More rationale for **rulers**:

Git commit messages have _two_ relevant column positions, for the subject line and the body text.
(https://cbea.ms/git-commit/ recommends "about 50" for the former, and 72 for the latter.)

So whenever I edit a git commit message, seeing rulers at columns 50 and 72 would make it much easier for me to conform to commit formatting conventions; I don't really need the editor to enforce anything, but I want to see if I'm overstepping.

Making it easier to conform to team standards does help with collaboration, so it looks to me that rulers are in scope of Editorconfig even if they are purely presentational.

**Comment by Timo-Weike** [comment]
Maybe this is also interesting for the disucssion.

[This ](https://marketplace.visualstudio.com/items?itemName=PaulHarrington.EditorGuidelines) is an extension for Visual Studio that uses the `.editorconfig` file to configure editor guidelines (aka rules).

It uses the property `guidelines` and also enables to some extend the styling of the rulers.

Example from my `.editorconfig` file
~~~
[*.cs]
guidelines = 80 1px dotted 30B0ED4C, 100 1px dotted ff0000
~~~

With this the Visual Studio draws a light grey dotted line after the 80th character and a dotted red line after the 100th character
in all `*.cs` files opend in the editor.

I'm currently pondering to develop an vs code extension to use the same property and syntax to also have my guideline/rulers in vs code.

**Comment by hakre** [comment]
@Timo-Weike This is absolutely amazing. If I may suggest for inspiration, and only if feasibly, it would be great to use a smaller/bigger narrowed/widened type-face in the editor before/after a ruler.

**Comment by JasinYip** [comment]
I can't believe it! This issue has been around for 10 years!🤣

**Comment by toolforger** [comment]
@hakre What would be your use case for different width after a ruler?

I am a bit worried about feature creep, raising the bar of implementing editorconfig for some editor.
Rulers are already quite a mouthful to implement as they are, and having too many features "that never get fully implemented everywhere" will immediately reduce editorconfig's usefulness in team environments.
Note that "usefulness to teams" is one of the reasons why editorconfig exists in the first place; I'd suggest prioritizing implementing missing features above defining new features.

**Comment by hakre** [comment]
@toolforger the use case is the same as with any rulers in editorconfig: visuals. and I'm totally against feature creep in implementations but totally for writing a specification to be portable and extensible. e.g. rulers should relate to CSS, check the current Snapshot on what is ready to implement. Never a must, always a can.

**Comment by toolforger** [comment]
@hakre
I wasn't talking about what the feature is about, I was talking about adding more features and the trade-offs involved.

If you say your suggestion is not feature creep, then you have to put forth arguments how the benefits of that feature increase the usefulness of editorconfig beyond the cost of having a higher entry barrier and more code to maintain.

**Comment by hakre** [comment]
Supporting the CSS nomenclature optionally per specification must and should not higher the barrier of entry nor result in more code to maintain. Not only with our extensively large rejection of feature creep in implementation, the mass of the whole solar system must appear minuscule in the light of its sun against it, we find it hard to travel this road of thought that lead to the opposite conclusions. 

> Um der Grazien willen, Peregrin, halt ein!
>
> Laß es an dieser Probe genug seyn!

@toolforger

**Comment by ssg** [comment]
I think that rendering styles shouldn't be part of editorconfig. I consider editorconfig as the way to share/enforce team standards during development, and rendering hints can't be part of that due to different preferences and accessibility requirements of developers.

Trying to define visual styles of things would be similar to specifying `tab_width` in pixels. What a pixel means changes dracstically between different developer machines. So, providing CSS styles in editorconfig just calls for many frustrated developers due to how their environment doesn't render that CSS properly. A good looking CSS on a light background would immediately fail on a dark theme for instance.

Guidelines themselves make sense to me because they're defined based on column numbers, so they're not subject to any inconsistencies between different developer machines. 

Related: https://github.com/editorconfig/editorconfig/issues/387#issuecomment-791937606

**Comment by alejandroiglesias** [comment]
This would be one of the most useful additions to EditorConfig. With modern tools such as Prettier to format code based on line lengths, we should also have the ability to display a ruler. It's a common editor config that is useful to share across a team when there are other tools also acting on line lengths (ESLint, Prettier, etc).

**Comment by hakre** [comment]
> This would be one of the most useful additions to EditorConfig. With modern tools such as Prettier to format code based on line lengths, we should also have the ability to display a ruler. It's a common editor config that is useful to share across a team when there are other tools also acting on line lengths (ESLint, Prettier, etc).

What's the current status of ESLint, Prettier, etc on editorconfig support? Aren't they extensible and read such a property out? How did this work for you? And if not, any references to downstream issues/patches? Just a friendly reminder, so perhaps if you find your idea is moving you, why not recuperate that energy and get things moving already today? Here in the PHP world we have a parse_ini_file() at hand, always, so things were already very easy to do back when this was originally reported within make recipes. IIRC Prettier is in the NodeJS context, but nevertheless they have scripts and I guess some ini-parser, too. If you prefix your stuff with an `example` namespace I see no reason to not give it a try. @alejandroiglesias

**Comment by alejandroiglesias** [comment]
@hakre they read certain properties for formatting, but not a property such as rulers or similar. In VSCode for example they need to give an API for extensions to configure rulers. A PR was in the works and then closed without merging: https://github.com/microsoft/vscode/pull/124879
For our team we solved it by adding a setting into the VSCode workspace settings that we can commit into the repo. I wouldn't have the time to look into this myself.

**Comment by ThiefMaster** [comment]
Could this be reopened on the vscode side? It was NOT closed by the VSCode maintainers but rather because the PR author deleted his vscode fork. It would be really nice to be able to set rulers via editorconfig in VSCode (and potentially other editors that support rulers). For example it's quite common to use different rulers for markdown and similar text files and source code.

**Comment by hakre** [comment]
Vscode maintains an open status, cf. https://github.com/microsoft/vscode/issues/111251 @ThiefMaster

**Comment by gphg** [comment]
It is 2026. I'll stick with the vscode's `editor.rules` from now on.
